pkg/helpers/tokencmd/request_token.go

```diff
@@ -397,39 +397,54 @@ func request(rt http.RoundTripper, requestURL string, requestHeaders http.Header
 	return rt.RoundTrip(req)
 }
 
+// transportWithSystemRoots tries to retrieve the serving certificate from the
+// issuer, validates it against the system roots and if the validation passes,
+// returns transport using just system roots, otherwise it returns a transport
+// that uses the CA from kubeconfig
 func transportWithSystemRoots(issuer string, clientConfig *restclient.Config) (http.RoundTripper, error) {
-	// copy the config so we can freely mutate it
-	configWithSystemRoots := restclient.CopyConfig(clientConfig)
-
-	// explicitly unset CA cert information
-	// this will make the transport use the system roots or OS specific verification
-	// this is required to have reasonable behavior on windows (cannot get system roots)
-	// in general there is no good with to say "I want system roots plus this CA bundle"
-	// so we just try system roots first before using the kubeconfig CA bundle
-	configWithSystemRoots.CAFile = ""
-	configWithSystemRoots.CAData = nil
-
-	systemRootsRT, err := restclient.TransportFor(configWithSystemRoots)
+	issuerURL, err := url.Parse(issuer)
 	if err != nil {
 		return nil, err
 	}
 
-	// build a request to probe the OAuth server CA
-	req, err := http.NewRequest(http.MethodHead, issuer, nil)
+	// perform the retrieval with insecure transport, otherwise oauth-server
+	// logs remote tls error which is confusing during troubleshooting
+	config := tls.Config{
+		Certificates:       []tls.Certificate{},
+		InsecureSkipVerify: true,
+		ServerName:         issuerURL.Hostname(),
+	}
+
+	port := issuerURL.Port()
+	if len(port) == 0 {
+		port = "443"
+	}
+	conn, err := tls.Dial("tcp", net.JoinHostPort(issuerURL.Hostname(), port), &config)
 	if err != nil {
 		return nil, err
 	}
+	conn.Close()
 
-	// see if get a certificate error when using the system roots
-	// we perform the check using this transport (instead of the kubeconfig based one)
-	// because it is most likely to work with a route (which is what the OAuth server uses in 4.0+)
-	// note that both transports are "safe" to use (in the sense that they have valid TLS configurations)
-	// thus the fallback case is not an "unsafe" operation
-	_, err = systemRootsRT.RoundTrip(req)
+	_, err = verifyServerCertChain(issuer, conn.ConnectionState().PeerCertificates)
```

Comment:
Is this supposed to be `issuerURL.Hostname()`? or `issuer` will work with this? Not sure if there is a difference between both.

Reply:
it seems that both work but hostname might actually be a better choice :+1: